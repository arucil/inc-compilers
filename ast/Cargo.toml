[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
