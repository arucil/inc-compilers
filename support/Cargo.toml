[package]
name = "support"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
assert_cmd = "2.0.4"
predicates = "2.1.1"
pretty_assertions = "1.2.1"
scopeguard = "1.1.0"
