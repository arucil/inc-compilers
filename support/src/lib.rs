pub mod compile_error;
pub mod range;
pub mod test_cli;

pub use self::compile_error::*;
pub use self::range::*;
pub use self::test_cli::*;
