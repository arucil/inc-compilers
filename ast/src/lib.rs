#![feature(never_type, box_syntax)]

use id_arena::{Arena, Id};
use pretty::*;
use std::fmt::{self, Debug, Formatter};
use std::iter;
use support::Range;

pub mod parser;

pub use parser::{parse, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<VAR = String, TYPE = ()> {
  pub body: Vec<Exp<VAR, TYPE>>,
  pub types: Arena<CompType>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Exp<VAR = String, TYPE = ()> {
  pub kind: ExpKind<VAR, TYPE>,
  pub range: Range,
  pub ty: TYPE,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExpKind<VAR, TYPE> {
  Int(i64),
  Prim {
    op: (Range, &'static str),
    args: Vec<Exp<VAR, TYPE>>,
  },
  Var(VAR),
  Str(String),
  Let {
    var: (Range, VAR),
    init: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Bool(bool),
  If {
    cond: Box<Exp<VAR, TYPE>>,
    conseq: Box<Exp<VAR, TYPE>>,
    alt: Box<Exp<VAR, TYPE>>,
  },
  Set {
    var: (Range, VAR),
    exp: Box<Exp<VAR, TYPE>>,
  },
  Get(VAR),
  Begin {
    seq: Vec<Exp<VAR, TYPE>>,
    last: Box<Exp<VAR, TYPE>>,
  },
  While {
    cond: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Void,
  Print(
    /// nonempty
    Vec<Exp<VAR, TYPE>>,
  ),
  NewLine,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Type {
  Int,
  Bool,
  Str,
  Comp(TypeId),
  Void,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TypeId(Id<CompType>);

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompType {
  Vector(Vec<Type>),
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct IdxVar {
  pub name: String,
  pub index: usize,
}

impl IdxVar {
  pub fn new<S: ToString>(name: S) -> Self {
    let name = name.to_string();
    if let Some(i) = name.find('.') {
      if let Ok(index) = name[i + 1..].parse::<usize>() {
        return Self {
          name: name[..i].to_owned(),
          index,
        };
      }
    }
    Self { name, index: 0 }
  }
}

impl Debug for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    write!(f, "{}.{}", self.name, self.index)
  }
}

impl<VAR: Debug, TYPE: Debug> Program<VAR, TYPE> {
  pub fn to_string_pretty(&self) -> String {
    let doc = self.to_doc();
    let mut buf = String::new();
    doc.render_fmt(30, &mut buf).unwrap();
    buf
  }

  fn to_doc(&self) -> RcDoc {
    RcDoc::intersperse(self.body.iter().map(|exp| exp.to_doc()), Doc::line())
  }
}

impl<VAR: Debug, TYPE: Debug> Exp<VAR, TYPE> {
  fn to_doc(&self) -> RcDoc {
    match &self.kind {
      ExpKind::Int(n) => RcDoc::text(format!("{}", n)),
      ExpKind::Str(s) => RcDoc::text(format!("{:?}", s)),
      ExpKind::Var(var) | ExpKind::Get(var) => {
        RcDoc::text(format!("{:?}", var))
      }
      ExpKind::Prim { op, args } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text(op.1))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Let { var, init, body } => RcDoc::text("(")
        .append(
          RcDoc::text("let")
            .append(Doc::line())
            .append(
              RcDoc::text("[")
                .append(
                  RcDoc::text(format!("{:?}", var.1))
                    .append(Doc::line())
                    .append(init.to_doc())
                    .nest(1)
                    .group(),
                )
                .append(RcDoc::text("]")),
            )
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Bool(true) => RcDoc::text("#t"),
      ExpKind::Bool(false) => RcDoc::text("#f"),
      ExpKind::If { cond, conseq, alt } => RcDoc::text("(")
        .append(
          RcDoc::text("if")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(conseq.to_doc().append(Doc::line()).append(alt.to_doc()))
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Set { var, exp } => RcDoc::text("(")
        .append(
          RcDoc::text("set!")
            .append(Doc::line())
            .append(RcDoc::text(format!("{:?}", var.1)))
            .append(Doc::line())
            .append(exp.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Begin { seq, last } => RcDoc::text("(")
        .append(
          RcDoc::text("begin").append(Doc::line()).append(
            RcDoc::intersperse(
              seq
                .iter()
                .chain(std::iter::once(&**last))
                .map(|exp| exp.to_doc()),
              Doc::line(),
            )
            .nest(1)
            .group(),
          ),
        )
        .append(RcDoc::text(")")),
      ExpKind::While { cond, body } => RcDoc::text("(")
        .append(
          RcDoc::text("while")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Void => RcDoc::text("(void)"),
      ExpKind::Print(args) => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text("print"))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::NewLine => RcDoc::text("(newline)"),
    }
  }
}
