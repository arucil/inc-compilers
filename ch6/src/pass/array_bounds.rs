use ast::{Error, Exp, ExpKind, FuncDef, IdxVar, Program, Type};
use id_arena::Arena;
use support::Range;

/// (make-vector e1 e2)
/// =>
/// (let ([len e1])
///   (if (>= len 0)
///     (make-vector len e2)
///     (length-error)))
///
/// (vector-ref e1 e2)
/// =>
/// (let ([vec e1] [idx e2])
///   (if (and (>= idx 0)
///            (< idx (vector-length vec)))
///     (vector-ref vec idx)
///     (out-of-bounds idx (vector-length vec))))
///
/// (vector-set! e1 e2 e3)
/// =>
/// (let ([vec e1] [idx e2])
///   (if (and (>= idx 0) (< idx (vector-length vec)))
///     (vector-set! vec idx e3)
///     (out-of-bounds idx (vector-length vec))))
pub fn insert_bounds_check(
  prog: Program<IdxVar, Type>,
) -> Program<IdxVar, Type> {
  let mut state = State {
    types: &prog.types,
    tmp_counter: 0,
  };
  Program {
    func_defs: prog
      .func_defs
      .into_iter()
      .map(|(name, func)| {
        (
          name,
          FuncDef {
            body: state.exp_insert(func.body),
            ..func
          },
        )
      })
      .collect(),
    body: prog
      .body
      .into_iter()
      .map(|exp| state.exp_insert(exp))
      .collect(),
    ..prog
  }
}

struct State<'a> {
  types: &'a Arena<Type>,
  tmp_counter: usize,
}

impl<'a> State<'a> {
  fn exp_insert(&mut self, exp: Exp<IdxVar, Type>) -> Exp<IdxVar, Type> {
    match exp.kind {
      ExpKind::Int(_)
      | ExpKind::Var(_)
      | ExpKind::Str(_)
      | ExpKind::Bool(_)
      | ExpKind::Get(_)
      | ExpKind::Void
      | ExpKind::NewLine => exp,
      ExpKind::Prim {
        op: op @ (_, "make-vector"),
        mut args,
      } => {
        let range = exp.range;
        let ty = exp.ty;
        let val = self.exp_insert(args.pop().unwrap());
        let len = self.exp_insert(args.pop().unwrap());
        self.make_check_exp(
          range,
          ty.clone(),
          vec![],
          len,
          |len_var| Exp {
            kind: ExpKind::Prim {
              op: (range, ">="),
              args: vec![
                len_var,
                Exp {
                  kind: ExpKind::Int(0),
                  range,
                  ty: Type::Int,
                },
              ],
            },
            range,
            ty: Type::Bool,
          },
          |len_var| Exp {
            kind: ExpKind::Prim {
              op,
              args: vec![len_var, val],
            },
            range,
            ty,
          },
          |len_var| Error::Length(box len_var),
        )
      }
      ExpKind::Prim {
        op: op @ (_, "vector-ref"),
        mut args,
      } if matches!(self.resolve_type(&args[0].ty), Type::Array(_)) => {
        let range = exp.range;
        let ty = exp.ty;
        let idx = self.exp_insert(args.pop().unwrap());
        let vec = self.exp_insert(args.pop().unwrap());
        let mut tmps = vec![];
        let vec_var;
        if vec.kind.is_atomic() {
          vec_var = vec;
        } else {
          let tmp = self.new_tmp();
          vec_var = Exp {
            kind: ExpKind::Var(tmp.clone()),
            range: vec.range,
            ty: vec.ty.clone(),
          };
          tmps.push(((vec.range, tmp), box vec));
        }
        self.make_check_exp(
          range,
          ty.clone(),
          tmps,
          idx,
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op: (range, "and"),
              args: vec![
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, ">="),
                    args: vec![
                      idx_var.clone(),
                      Exp {
                        kind: ExpKind::Int(0),
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, "<"),
                    args: vec![
                      idx_var,
                      Exp {
                        kind: ExpKind::Prim {
                          op: (range, "vector-length"),
                          args: vec![vec_var.clone()],
                        },
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
              ],
            },
            range,
            ty: Type::Bool,
          },
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op,
              args: vec![vec_var.clone(), idx_var],
            },
            range,
            ty: ty.clone(),
          },
          |idx_var| Error::OutOfBounds {
            index: box idx_var,
            len: box Exp {
              kind: ExpKind::Prim {
                op: (range, "vector-length"),
                args: vec![vec_var.clone()],
              },
              range,
              ty: Type::Int,
            },
          },
        )
      }
      ExpKind::Prim {
        op: op @ (_, "vector-set!"),
        mut args,
      } if matches!(self.resolve_type(&args[0].ty), Type::Array(_)) => {
        let range = exp.range;
        let ty = exp.ty;
        let val = self.exp_insert(args.pop().unwrap());
        let idx = self.exp_insert(args.pop().unwrap());
        let vec = self.exp_insert(args.pop().unwrap());
        let mut tmps = vec![];
        let vec_var;
        if vec.kind.is_atomic() {
          vec_var = vec;
        } else {
          let tmp = self.new_tmp();
          vec_var = Exp {
            kind: ExpKind::Var(tmp.clone()),
            range: vec.range,
            ty: vec.ty.clone(),
          };
          tmps.push(((vec.range, tmp), box vec));
        }
        self.make_check_exp(
          range,
          ty.clone(),
          tmps,
          idx,
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op: (range, "and"),
              args: vec![
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, ">="),
                    args: vec![
                      idx_var.clone(),
                      Exp {
                        kind: ExpKind::Int(0),
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
                Exp {
                  kind: ExpKind::Prim {
                    op: (range, "<"),
                    args: vec![
                      idx_var,
                      Exp {
                        kind: ExpKind::Prim {
                          op: (range, "vector-length"),
                          args: vec![vec_var.clone()],
                        },
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range,
                  ty: Type::Bool,
                },
              ],
            },
            range,
            ty: Type::Bool,
          },
          |idx_var| Exp {
            kind: ExpKind::Prim {
              op,
              args: vec![vec_var.clone(), idx_var, val],
            },
            range,
            ty: ty.clone(),
          },
          |idx_var| Error::OutOfBounds {
            index: box idx_var,
            len: box Exp {
              kind: ExpKind::Prim {
                op: (range, "vector-length"),
                args: vec![vec_var.clone()],
              },
              range,
              ty: Type::Int,
            },
          },
        )
      }
      ExpKind::Prim { op, args } => Exp {
        kind: ExpKind::Prim {
          op,
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        },
        ..exp
      },
      ExpKind::Apply {
        func,
        args,
        r#struct,
      } => Exp {
        kind: ExpKind::Apply {
          func: box self.exp_insert(*func),
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          r#struct,
        },
        ..exp
      },
      ExpKind::Let { var, init, body } => Exp {
        kind: ExpKind::Let {
          var,
          init: box self.exp_insert(*init),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::If { cond, conseq, alt } => Exp {
        kind: ExpKind::If {
          cond: box self.exp_insert(*cond),
          conseq: box self.exp_insert(*conseq),
          alt: box self.exp_insert(*alt),
        },
        ..exp
      },
      ExpKind::Set { var, exp: exp1 } => Exp {
        kind: ExpKind::Set {
          var,
          exp: box self.exp_insert(*exp1),
        },
        ..exp
      },
      ExpKind::Begin { seq, last } => Exp {
        kind: ExpKind::Begin {
          seq: seq.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          last: box self.exp_insert(*last),
        },
        ..exp
      },
      ExpKind::While { cond, body } => Exp {
        kind: ExpKind::While {
          cond: box self.exp_insert(*cond),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::Print(args) => Exp {
        kind: ExpKind::Print(
          args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        ),
        ..exp
      },
      ExpKind::Error(_) => unreachable!(),
      ExpKind::FunRef { .. } => exp,
    }
  }

  fn make_check_exp<F, G, E>(
    &mut self,
    range: Range,
    ty: Type,
    mut tmps: Vec<Tmp>,
    index: Exp<IdxVar, Type>,
    cond: F,
    conseq: G,
    err: E,
  ) -> Exp<IdxVar, Type>
  where
    F: FnOnce(Exp<IdxVar, Type>) -> Exp<IdxVar, Type>,
    G: FnOnce(Exp<IdxVar, Type>) -> Exp<IdxVar, Type>,
    E: FnOnce(Exp<IdxVar, Type>) -> Error<IdxVar, Type>,
  {
    let idx_var;
    if index.kind.is_atomic() {
      idx_var = index;
    } else {
      let tmp = self.new_tmp();
      idx_var = Exp {
        kind: ExpKind::Var(tmp.clone()),
        range: index.range,
        ty: Type::Int,
      };
      tmps.push(((index.range, tmp), box index));
    }
    let body = Exp {
      kind: ExpKind::If {
        cond: box cond(idx_var.clone()),
        conseq: box conseq(idx_var.clone()),
        alt: box Exp {
          kind: ExpKind::Error(err(idx_var)),
          range,
          ty: ty.clone(),
        },
      },
      range,
      ty: ty.clone(),
    };
    tmps.into_iter().rfold(body, |body, (var, init)| Exp {
      kind: ExpKind::Let {
        var,
        init,
        body: box body,
      },
      range,
      ty: ty.clone(),
    })
  }

  fn resolve_type(&self, ty: &Type) -> Type {
    ty.resolved(self.types)
  }

  fn new_tmp(&mut self) -> IdxVar {
    self.tmp_counter += 1;
    IdxVar {
      name: "(bounds-tmp)".to_owned(),
      index: self.tmp_counter,
    }
  }
}

type Tmp = ((Range, IdxVar), Box<Exp<IdxVar, Type>>);

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn array() {
    let prog = parse(
      r#"
(let ([x (make-vector 3 #t)]
      [i 1]
      [y (vector-ref x i)]
      [z (vector 1 #t "abc")])
  (vector-set! x i (vector-ref z 1))
  (vector-set! z 1 (not y)))
      "#,
    )
    .unwrap();
    let prog = typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let result = insert_bounds_check(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
