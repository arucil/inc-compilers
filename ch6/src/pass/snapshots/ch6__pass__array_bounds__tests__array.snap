---
source: ch6/src/pass/array_bounds.rs
expression: result.to_string_pretty()
---
(let
 [x
  (let
   [(bounds-tmp-index) 3]
   (if
    (>= (bounds-tmp-index) 0)
    (make-vector
     (bounds-tmp-index)
     #t)
    (length-error
     (bounds-tmp-index))))]
 (let
  [i 1]
  (let
   [y
    (let
     [(bounds-tmp-vec) x]
     (let
      [(bounds-tmp-index) i]
      (if
       (and
        (>=
         (bounds-tmp-index)
         0)
        (<
         (bounds-tmp-index)
         (vector-length
          (bounds-tmp-vec))))
       (vector-ref
        (bounds-tmp-vec)
        (bounds-tmp-index))
       (out-of-bounds-error
        (bounds-tmp-index)
        (vector-length
         (bounds-tmp-vec))))))]
   (let
    [z (vector 1 #t "abc")]
    (begin
    (let
      [(bounds-tmp-vec) x]
      (let
       [(bounds-tmp-index) i]
       (if
        (and
         (>=
          (bounds-tmp-index)
          0)
         (<
          (bounds-tmp-index)
          (vector-length
           (bounds-tmp-vec))))
        (vector-set!
         (bounds-tmp-vec)
         (bounds-tmp-index)
         (vector-ref z 1))
        (out-of-bounds-error
         (bounds-tmp-index)
         (vector-length
          (bounds-tmp-vec))))))
     (vector-set!
      z
      1
      (not y)))))))
